src/core/util/tag.h

```diff
@@ -56,8 +56,10 @@ class Tag
 
   Tag() : m_Buf{} {}
 
-  Tag(const std::uint8_t* buf)
+  Tag(const std::uint8_t* buf) : m_Buf{}
```

Comment:
>Tag copy-constructor needs non-null buffers for well-defined memcpy.

...but this isn't a copy-constructor. What compiler are you using?

> : m_Buf{}

I intentionally avoided this unnecessary initialization in https://github.com/monero-project/kovri/commit/507b888f337161d52c4e28626180fd2deac3cc38 because, with memcpy, so long as the source is not null, you'll be copying over god-knows-what anyway; and the size of the buffer is templated so there's no uncertainty of the length copied.

Reply:
> this isn't a copy-constructor

You're right, I was using the wrong term.

> I intentionally avoided this unnecessary initialization in 507b888 because, with memcpy, so long as the source is not null, you'll be copying over god-knows-what anyway; and the size of the buffer is templated so there's no uncertainty of the length copied.

I understand part of your reasoning, length is guaranteed safe. However, isn't writing over uninitialized memory undefined behavior (and potentially dangerous)? Sincerely asking, since I might be misunderstanding something.

Reply:
I think you're confusing idioms, and confusing allocation with initialization. The destination memory is allocated before the copy. As for undefined behavior, it is not undefined because the destination is not a nullptr and it is also TriviallyCopyable.

So, to clarify: upon construction, zero-initializing allocated memory with `Size` only to subsequently fill it with an unknown source of `Size`, would be considered an "unnecessary initialization". Most importantly, since we're not pulling anything from the buffer before the copy, there shouldn't be concern about pulling from uninitialized memory (which would definitely be a cause for concern). 

Does that make sense?

Reply:
Makes complete sense now, and I appreciate you explaining it. I didn't understand the part about the destination buffer being allocated before the copy. Was mistakenly thinking that the buffer would be pointing to some arbitrary location on the heap that could be occupied by something else. Now, I understand that it has its own heap location, and that location is filled with whatever garbage.